Investigate CI flakiness for repo-updater/repos/sources_test.go

We had to run our build two times to make it pass on CI: https://buildkite.com/sourcegraph/sourcegraph/builds/34892#0d29f393-5cd6-4406-84f5-2723c5d47752

```

FAIL: TestGithubSource_GetRepo (0.02s) | 54s
-- | --
  | [2019-05-14T14:31:11Z]     --- FAIL: TestGithubSource_GetRepo/GITHUB-DOT-COM/not_found (0.00s)
  | [2019-05-14T14:31:11Z]         sources_test.go:857: error:
  | [2019-05-14T14:31:11Z]             have: "GitHub repository not found"
  | [2019-05-14T14:31:11Z]             want: "request to http://github-proxy/repos/foobarfoobarfoobar/please-let-this-not-exist returned status 404: Not Found"
  | [2019-05-14T14:31:11Z] FAIL

```

The same test didn't fail in previous PR and it doesn't fail locally. Our best guess at the moment is that this has something to do with the caching logic in the Github client:

https://github.com/sourcegraph/sourcegraph/blob/d6294814e1102a871eb28669f960997434904baf/pkg/extsvc/github/repos.go#L139-L145

This is also related to #3895

--- a/cmd/repo-updater/repos/sources_test.go
+++ b/cmd/repo-updater/repos/sources_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/pkg/extsvc/phabricator"
 	"github.com/sourcegraph/sourcegraph/pkg/httpcli"
 	"github.com/sourcegraph/sourcegraph/pkg/httptestutil"
+	"github.com/sourcegraph/sourcegraph/pkg/rcache"
 	"github.com/sourcegraph/sourcegraph/schema"
 	log15 "gopkg.in/inconshreveable/log15.v2"
 )
@@ -833,7 +834,13 @@ func TestGithubSource_GetRepo(t *testing.T) {
 	for _, tc := range testCases {
 		tc := tc
 		tc.name = "GITHUB-DOT-COM/" + tc.name
+
 		t.Run(tc.name, func(t *testing.T) {
+			// The GithubSource uses the github.Client under the hood, which
+			// uses rcache, a caching layer that uses Redis.
+			// We need to clear the cache before we run the tests
+			rcache.SetupForTest(t)
+
 			cf, save := newClientFactory(t, tc.name)
 			defer save(t)
 

--- a/cmd/repo-updater/repos/sources_test.go
+++ b/cmd/repo-updater/repos/sources_test.go
@@ -790,7 +790,7 @@ func TestGithubSource_GetRepo(t *testing.T) {
 		{
 			name:          "not found",
 			nameWithOwner: "foobarfoobarfoobar/please-let-this-not-exist",
-			err:           `request to http://github-proxy/repos/foobarfoobarfoobar/please-let-this-not-exist returned status 404: Not Found`,
+			err:           `GitHub repository not found`,
 		},
 		{
 			name:          "found",

--- a/pkg/extsvc/github/repos.go
+++ b/pkg/extsvc/github/repos.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"net/http"
 	"net/url"
 	"strconv"
 	"strings"
@@ -252,6 +253,9 @@ func (c *Client) getRepositoryFromAPI(ctx context.Context, owner, name string) (
 	// http://[sourcegraph-hostname]/github.com/foo/bar.
 	var result restRepository
 	if err := c.requestGet(ctx, "", fmt.Sprintf("/repos/%s/%s", owner, name), &result); err != nil {
+		if HTTPErrorCode(err) == http.StatusNotFound {
+			return nil, ErrNotFound
+		}
 		return nil, err
 	}
 	return convertRestRepo(result), nil

--- a/pkg/extsvc/github/repos_test.go
+++ b/pkg/extsvc/github/repos_test.go
@@ -140,6 +140,9 @@ func TestClient_GetRepository_nonexistent(t *testing.T) {
 	if !IsNotFound(err) {
 		t.Errorf("got err == %v, want IsNotFound(err) == true", err)
 	}
+	if err != ErrNotFound {
+		t.Errorf("got err == %q, want ErrNotFound", err)
+	}
 	if repo != nil {
 		t.Error("repo != nil")
 	}

